framework/src/base/MooseApp.C

```diff
@@ -2083,112 +2089,101 @@ MooseApp::attachRelationshipManagers(Moose::RelationshipManagerType rm_type,
-
-          problem.mesh().getMesh().add_ghosting_functor(*rm);
+          problem.getDisplacedProblem()->addAlgebraicGhostingFunctor(clone_gf,
+                                                                     /*to_mesh = */ false);
+          _undisp_to_disp_rms.emplace(rm.get(), clone_gf);
         }
+      }
+      else // undisplaced
+      {
+        if (rm_type == Moose::RelationshipManagerType::COUPLING)
+          undisp_nl_dof_map.add_coupling_functor(*rm, /*to_mesh = */ false);
 
-        if (rm->useDisplacedMesh() && problem.getDisplacedProblem())
-        {
-          if (rm_type == Moose::RelationshipManagerType::COUPLING)
-          {
-            // We actually need to add this to the FEProblemBase NonlinearSystemBase's DofMap
-            // because the DisplacedProblem "nonlinear" DisplacedSystem doesn't have any matrices
-            // for which to do coupling
-            auto & dof_map = problem.getNonlinearSystemBase().dofMap();
-            dof_map.add_coupling_functor(*rm, /*to_mesh = */ false);
-            rm->setDofMap(dof_map);
-          }
-          // If this rm is algebraic AND coupling, then in the case of the non-linear system there
-          // is no reason to add it to the DofMap twice. In the case of any other system, it
-          // actually would be disastrous to add this rm because it's going to set a coupling matrix
-          // based on the non-linear system. So we don't add this rm at all here if its also
-          // a coupling functor
-          else if (rm_type == Moose::RelationshipManagerType::ALGEBRAIC &&
-                   !rm->isType(Moose::RelationshipManagerType::COUPLING))
-            problem.getDisplacedProblem()->addAlgebraicGhostingFunctor(clone_rm,
-                                                                       /*to_mesh = */ false);
-        }
-        else // undisplaced
-        {
-          if (rm_type == Moose::RelationshipManagerType::COUPLING)
-          {
-            auto & dof_map = problem.getNonlinearSystemBase().dofMap();
-            dof_map.add_coupling_functor(*rm, /*to_mesh = */ false);
-            rm->setDofMap(dof_map);
-          }
-          // If this rm is algebraic AND coupling, then in the case of the non-linear system there
-          // is no reason to add it to the DofMap twice. In the case of any other system, it
-          // actually would be disastrous to add this rm because it's going to set a coupling matrix
-          // based on the non-linear system. So we don't add this rm at all here if its also
-          // a coupling functor
-          else if (rm_type == Moose::RelationshipManagerType::ALGEBRAIC &&
-                   !rm->isType(Moose::RelationshipManagerType::COUPLING))
-            problem.addAlgebraicGhostingFunctor(*rm, /*to_mesh = */ false);
-        }
+        else if (rm_type == Moose::RelationshipManagerType::ALGEBRAIC)
+          problem.addAlgebraicGhostingFunctor(*rm, /*to_mesh = */ false);
```

Comment:
It is a great refactor of `MooseApp::attachRelationshipManagers`. The logic is so simple and also very easy to understand